### ISSUE TITLE
The loss computation with `size_average` should average over batch example or batch element ?

### ISSUE THREAD

**Issue by zuoxingdong** [pytorch/pytorch#6988]
Traditionally, when we have a batched data with shape `(N, D)` where `N` is batch size and `D` is data dimension. The losses are often calculated for each training example say
```
L_i = loss(X_i), i = 1, ..., N
```
And then total loss is averaged over the batch size
```
L = (1/N)*sum(L_i)
```

However, it seems it is not what `nn.*Loss` is doing for the flag `size_average=True/False`

e.g. 

```python
import torch
import torch.nn.functional as F

input = torch.randn(3, 2)
target = torch.rand(3, 2)

print(input)
print(target)

full_loss = F.mse_loss(input, target, reduce=False)
print(full_loss)
loss_sum = F.mse_loss(input, target, size_average=False)
loss_mean = F.mse_loss(input, target, size_average=True)
print(loss_sum)
print(loss_mean)

batch_loss = full_loss.sum(dim=-1)
correct_loss_mean = batch_loss.mean()
correct_loss_sum = batch_loss.sum()
print(correct_loss_mean)
print(correct_loss_sum)

assert correct_loss_sum == loss_sum
assert correct_loss_mean == loss_mean
```
it seems the `size_average` does not properly average the loss over the batch of examples, but averaged over all dimensions.

**Comment by zou3519** [comment]
`size_average` averages over "each atomic element for which loss is computed for". For `mse_loss` `size_average` divides by _all_ elements. For something like `NLLLoss`, `size_average` divides by number of minibatches (`tensor.size(0)`) because each row in the tensor results in a loss.

We'll definitely make the `size_average` behavior clearer in the docs: this question has come up in a number of issues in the past. If you'd like different behavior, you can set `size_average=False` and divide by the number of batches.

**Comment by zuoxingdong** [comment]
@zou3519 Thanks a lot for the explanation. Just to make sure I don't take it wrong. By setting `size_average=True` by default to average over all elements rather than firstly summing up losses for each training example and then to average. This does not alter the gradient direction but indeed a smaller scale of the gradient. Will it have potential effects to slow down the training ? Because averaging over all elements gets smaller loss, leading to smaller gradient norm.

**Comment by fmassa** [comment]
We should at some point consider if we want to change this behavior. This is indeed confusing and in some cases doesn't do what you want (KL Divergence for example https://github.com/pytorch/pytorch/pull/7006).
Breaking everyones code is not a good thing, but maybe we could deprecate the `size_average` argument in favor of a new one, that consistently divides by the batch size?

**Comment by apaszke** [comment]
@fmassa I'm a bit afraid that adding another variant of `size_average` is going to make the interface very confusing.

**Comment by fmassa** [comment]
@apaszke I think it might be worth listing the cases where `size_average` is actually doing what most people want it to do. Maybe the current behavior of `size_average` is already good enough.

That being said, now that we have `reduce=False` for all losses (and once we get reduction over multiple dimensions merged in master), it will be super easy to let users control what they want to do.

**Comment by zuoxingdong** [comment]
@fmassa It seems to me that the more common behavior of `size_average` should be firstly summing up the losses across data dimension and then averaging over batch. But yes, as you pointed out, the flag `reduce=False` provides user full flexibility. 

I guess it might be good either to mention clearly to the user about the current behavior of `size_average` is to average over all elements, or change the behavior of `size_average` to the one described above which is more commonly used to train NN models and thus more user-friendly, and then the 'element-wise' averaging behavior leave to the user. 

@apaszke 
Or maybe have one more flag, say `size_average_batch` and `size_average_all` to provide both behaviors.

**Comment by li-roy** [comment]
Does it make more sense to average over batch always? It seems like that's the functionality that most users expect and want.
